arch/arm/core/thread.c

```diff
@@ -91,10 +91,10 @@ void z_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 	 * k_thread_create(). If K_THREAD_STACK_SIZEOF() is used, the
 	 * Guard size has already been take out of stackSize.
 	 */
-	stackEnd = pStackMem + stackSize - MPU_GUARD_ALIGN_AND_SIZE;
-#else
-	stackEnd = pStackMem + stackSize;
+	stackSize -= MPU_GUARD_ALIGN_AND_SIZE;
 #endif
+	stackEnd = pStackMem + stackSize;
+
 	struct __esf *pInitCtx;
 
 	_new_thread_init(thread, pStackMem, stackSize, priority,
```

Comment:
Hm, I think I know why this was missed; because initially I wanted to skip the workaround for power-of-2-alignment. The change was introduced here: a23a542b801235fc00bec7bae8d465e105f40f9a

Could you instead revert this and simply call _new_thread_init with stackEnd - stackMem?

Reply:
> Could you instead revert this and simply call _new_thread_init with stackEnd - stackMem?

Why is that preferable? stackSize is simply *wrong*

Reply:
I am still not sure this is a bug - let me take a closer look

Reply:
Oh it's a bug. You have, in the thread struct, a stack bounds setting which ends 32 bytes past the stack object!

Reply:
See https://github.com/zephyrproject-rtos/zephyr/pull/13650

Reply:
I am not convinced this is a bug; at least, not an active bug. The reason is simple; the `stack_info.size` needs to have the nominal size of the stack object, in order to put the whole area into the "user" MPU region. If you check the thread stack programming in `arm_core_mpu.c`, you will see that the MPU boundaries of the thread stack are `base = stack_obj `(not `stack_info.start`) and `stack_info.size`.

Reply:
It is a bug.

The stack info in the struct k_thread has the wrong boundaries. This information is supposed to be correct and it gets used in various places.

Reply:
```
starting test - test_stack_buffer
Reserved space: 0
Provided stack size: 1024
 - Stack object 0x20004800[1024]
 - Testing supervisor mode
   - Thread reports buffer 0x20004820 size 1024

    Assertion failed at /projects/zephyr2/zephyr/tests/kernel/mem_protect/userspace/src/main.c:982: stack_buffer_scenarios: (stack_end <= obj_end is false)
stack size in thread struct out of bounds (overflow)
FAIL - test_stack_buffer

```

Reply:
OK I see, I was focusing only on the MPU; there is clearly no bug there.

However, if this patch gets in, we need a fix in the MPU re-programming (on how to use the boundaries for the thread stack). I can take that over.